lib/posix/pthread_key.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <posix/pthread.h>
+
+struct k_sem pthread_key_sem;
+
+typedef struct pthread_thread_data {
+	sys_dnode_t node;
+
+	/* Key and thread specific data passed to pthread_setspecific() */
+	pthread_key_obj *key;
+	void *spec_data;
+
+} pthread_thread_data;
+
+typedef struct pthread_key_data {
+	sys_dnode_t node;
+	pthread_thread_data thread_data;
+} pthread_key_data;
+
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *new_key;
+
+	*key = NULL;
+
+	k_sem_init(&pthread_key_sem, 1, 1);
+
+	new_key = k_malloc(sizeof(pthread_key_obj));
+
+	if (new_key == NULL) {
+		return ENOMEM;
+	}
+
+	sys_dlist_init(&(new_key->key_data_l));
+
+	new_key->destructor = destructor;
+	*key = (void *)new_key;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	pthread_key_data *key_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	/* Delete thread-specific elements associated with the key */
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(key_obj->key_data_l),
+			node_l, next_node_l) {
+
+		/* Remove the object from the list key_data_l */
+		key_data = (pthread_key_data *)
+			sys_dlist_get(&(key_obj->key_data_l));
+
+		/* Deallocate the object's memory */
+		k_free((void *)key_data);
+	}
+
+	k_free((void *)key_obj);
+
+	k_sem_give(&pthread_key_sem);
+
+	return 0;
+}
+
+/**
+ * @brief Associate a thread-specific value with a key
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setspecific(pthread_key_t key, const void *value)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	struct posix_thread *thread = (struct posix_thread *)pthread_self();
+	pthread_key_data *key_data;
+	pthread_thread_data *thread_spec_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	/* Traverse the list of keys set by the thread, looking for key.
+	 * If the key is already in the list, re-assign its value.
+	 * Else add the key to the thread's list.
+	 */
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(thread->key_list), node_l, next_node_l) {
+
+			thread_spec_data = (pthread_thread_data *)node_l;
+
+			if (thread_spec_data->key == key_obj) {
+
+				/* Key is already present so
+				 * associate thread specific data
+				 */
+				thread_spec_data->spec_data = (void *)value;
+				goto out;
+			}
+	}
+
+	if (node_l == NULL) {
+		key_data = k_malloc(sizeof(pthread_key_data));
+
+		if (key_data == NULL) {
+			return ENOMEM;
```

Comment:
pthread_key_sem is locked here.  Create a `retval` variable, set it to `ENOMEM`, and `goto out`.

Reply:
oh yeah, fixed this now